### DISCUSSION TITLE
Light weight Nvidia docker images

### DISCUSSION THREAD

**Discussion by agunapal** [moby/moby#46426] [Ideas]
### Description

It would be great to have light weight NVIDIA docker images.

Example: 
- The compressed size of [Nvidia CUDA 11.8 developer docker image](https://hub.docker.com/layers/nvidia/cuda/11.8.0-devel-ubuntu20.04/images/sha256-b7d90766e2fdb69fa30dcd04a001b25ab57d833dd155045ebf8982205fc1fca1?context=explore) is 2.92 GB.
 - The uncompressed size is  7.37 GB
```
nvidia/cuda          11.8.0-devel-ubuntu20.04          ff3049f3773d   2 months ago   7.37GB
```

To support state of the AI libraries, the developer image may be needed. 

Wondering if there is a common request yet from docker community - To have a light weight/slim version of these images?
With the explosion of generative AI use-cases, I wonder if it makes sense for the community to think about this. This could have a big impact on cost savings

In my research, I have seen docker enthusiasts trying to build a solution using slim as the base image, but I couldn't come across a good solution.

Wanted to get your thoughts on this request

**Comment by thaJeztah** [comment]
I don't think this is the right place to ask; that image is maintained by NVIDIA (but they happen to use Docker Hub to distribute the image); from the information on Docker Hub, it looks like the source-repository is https://gitlab.com/nvidia/cuda, so that may be the best location to ask the maintainers of that image;

<img width="828" alt="Screenshot 2023-09-08 at 01 41 00" src="https://github.com/moby/moby/assets/1804568/3697a372-693d-4ce5-8096-f9f03f307eae">

**Comment by chauhang** [comment]
@thaJeztah We are from the PyTorch team and looking for ways to improve the docker footprint for PyTorch and related libraries installs. We would like to partner with Docker to find a better solution. Current images from Nvidia have lot of bloat making it challenging to use esp for Inference workloads. Every customer has to go through the pain point of creating optimized images for their environment. Are there any other suggestions from Docker development team that we can explore? Will be happy to discuss more in a meeting

**Reply by polarathene** [reply]
I have a bit of experience with optimizing image builds. Here's some tips:
- Use a tool like [`dive`](https://github.com/wagoodman/dive) to get insights on an images layers and where the bulk of the size is introduced.
- You can use `RUN --mount=type=`
  - with type `cache` for package installs as a nice way to keep this extra weight out of an image, and retain the package manager cache for future builds should the image layer be invalidated in a future build (_speeds up this part when the cache can be reused_).
  - with type `bind` when you need to only have some data temporarily during build. This can be used to bring in files from other Docker images or earlier build stages instead of a `COPY`, avoiding redundant image weight.
    - [Here is an example](https://github.com/oven-sh/bun/pull/6100/files#diff-bf50d11c39869a68a22a512730cb43dc5e98e561f4204a2aedb78f20e2d1a9daR63-R67) I contributed before, symlinks could not be copied directly (_they'd be resolved to another copy of the file_) so the needed binaries are accessed from an earlier image stage that had them to create the symlink allowing for a much slimmer image.
- Use the HereDoc feature to avoid `&& \` noise for multi-line `RUN` or many separate `RUN` as the nvidia `Dockerfile` has (_that one isn't related to any of the DockerHub images though_).
- I gave some [informal review tips / advice on improving a `Dockerfile`](https://github.com/rspamd/rspamd/issues/4400#issuecomment-1743918884) which may be helpful (_[compression estimate](https://github.com/rspamd/rspamd/issues/4400#issuecomment-1744857889), [avoiding `VOLUME`](https://github.com/rspamd/rspamd/issues/4400#issuecomment-1745881053), and [build-time only `ENV` via `ARG`](https://github.com/rspamd/rspamd/issues/4400#issuecomment-1745881110)_).
- Using a base image like Fedora or openSUSE there is support for configuring an "install root", which will allow you to specify only the packages you need and create a whole new base image tailored to only supporting that.
  - [Here's an example](https://github.com/drwetter/testssl.sh/pull/2344/commits/1392987f01b748021cda4e98cbaf8e151c2f9a32#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R14-R66) with openSUSE and `--installroot`, it also uses the `RUN --mount=type=cache` and HereDoc features for you to reference. After that single `RUN` layer, a new build stage in the image is introduced to [transfer that install root location to it's own image](https://github.com/drwetter/testssl.sh/pull/2344/commits/1392987f01b748021cda4e98cbaf8e151c2f9a32#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R69-R72).
  - You may want to try the "install root" optimization last, as it doesn't always go smoothly. Personally I found [openSUSE had compatibility issues](https://github.com/openSUSE/libzypp/issues/444) within the container environment. Fedora was a better experience, but I don't have a reference to link you to for that.

DockerHub images (_12.2.0 - Ubuntu 22.04_):
- [Base](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.2.0/ubuntu2204/base/Dockerfile?ref_type=heads)
- [Runtime](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.2.0/ubuntu2204/runtime/Dockerfile?ref_type=heads)
- [Devel](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.2.0/ubuntu2204/devel/Dockerfile?ref_type=heads)

The bulk of the runtime weight is from a single layer that installs the runtime cuda libs, there's not much you can do about that AFAIK (_unless you know what you can remove_):

![image](https://github.com/moby/moby/assets/5098581/02fdf5fb-2575-4e7f-a40f-c1e48219b1ed)

I was not able to run `dive` on the `devel` variant as this consumed all 17GB of disk I had remaining (_WSL2 with `docker pull` command_), and failed. Part of the issue there is WSL2 memory usage for the file buffer / cache seems to also use up actual disk space in addition to memory.

Presumably there isn't really much that can be saved at a glance. Bulk of the weight is from the package installs, so it's unclear what portion of that is "bloat" as I guess it'd vary by project?

If you already have identified optimized images being created for environments, perhaps you could document those and find a common optimized base with additional advice to customize for other needs?

**Comment by neersighted** [comment]
If you're interested in working with Docker on base images, the place to ask/make suggestions is probably [docker-library/official-images](https://github.com/docker-library/official-images).

That being said, generally images there are required to be free software (and blessed or directly maintained by the 'upstream,' in this case NVIDIA), which means that this may be out of scope for the project. In that case, you would be best off partnering with NVIDIA to create more suitable base images/improve their images, as DOI/docker-library is the only Docker Inc. sponsored image-authoring project/process.
